pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -472,20 +472,93 @@ class StateVectorKokkos final
         PL_ABORT_IF_NOT(controlled_wires.size() == controlled_values.size(),
                         "`controlled_wires` must have the same size as "
                         "`controlled_values`.");
-        PL_ABORT_IF_NOT(
-            array_contains(controlled_gate_names, std::string_view{opName}),
-            "Controlled matrix operation not yet supported.");
 
         if (controlled_wires.empty()) {
             return applyOperation(opName, wires, inverse, params, gate_matrix);
         }
+        if (array_contains(controlled_gate_names, std::string_view{opName})) {
+            const std::size_t num_qubits = this->getNumQubits();
+            const ControlledGateOperation gateop =
+                reverse_lookup(controlled_gate_names, std::string_view{opName});
+            applyNCNamedOperation<KokkosExecSpace>(
+                gateop, *data_, num_qubits, controlled_wires, controlled_values,
+                wires, inverse, params);
+        } else {
+            PL_ABORT_IF(gate_matrix.empty(),
+                        std::string("Operation does not exist for ") + opName +
+                            std::string(" and no matrix provided."));
+            return applyNCMultiQubitOp(vector2view(gate_matrix),
+                                       controlled_wires, controlled_values,
+                                       wires, inverse);
+        }
+    }
 
-        const std::size_t num_qubits = this->getNumQubits();
-        const ControlledGateOperation gateop =
-            reverse_lookup(controlled_gate_names, std::string_view{opName});
-        applyNCNamedOperation<KokkosExecSpace>(
-            gateop, *data_, num_qubits, controlled_wires, controlled_values,
-            wires, inverse, params);
+    /**
+     * @brief Apply a controlled-multi qubit operator to the state vector using
+     * a matrix
+     *
+     * @param matrix Kokkos gate matrix in the device space.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (true or false).
+     * @param wires Wires to apply gate to.
+     * @param inverse Indicates whether to use adjoint of gate. (Default to
+     * false)
+     */
+    void applyNCMultiQubitOp(const KokkosVector matrix,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             bool inverse = false) {
+
+        auto &&num_qubits = this->getNumQubits();
+        std::size_t two2N =
+            std::exp2(num_qubits - wires.size() - controlled_wires.size());
```

Comment:
What about using `Pennylane::Util::exp2`?

Reply:
can we mark two2N as a const?

Reply:
I believe that's the idea behind [Util::exp2](https://github.com/PennyLaneAI/pennylane-lightning/blob/d473c6116076ea2b428633636aeb0e7ab25821fa/pennylane_lightning/core/src/utils/Util.hpp#L240). Could you ensure that we use this exp2 instead of `std::exp2`. I think you used it in a few other places in your previous PRs.

Reply:
Thanks, I'll update to `Util::exp2`